patches/minecraft/net/minecraft/server/packs/VanillaPackResources.java.patch

```diff
@@ -1,5 +1,63 @@
 --- a/net/minecraft/server/packs/VanillaPackResources.java
 +++ b/net/minecraft/server/packs/VanillaPackResources.java
+@@ -40,6 +_,26 @@
+    public static Path f_10312_;
+    private static final Logger f_10315_ = LogUtils.getLogger();
+    public static Class<?> f_10313_;
++   private final net.minecraftforge.resource.ResourceCacheManager cacheManager = new net.minecraftforge.resource.ResourceCacheManager(() -> false, VanillaPackResources::shouldIndexOffThread, (packType, namespace) -> f_182296_.get(packType).resolve(namespace));
++
++   private static boolean shouldIndexOffThread() {
++      return !getConfigValue(net.minecraftforge.common.ForgeConfig.COMMON.indexVanillaPackCachesOnThread);
++   }
++
++   private static boolean shouldUseCache() {
++      return getConfigValue(net.minecraftforge.common.ForgeConfig.COMMON.cachePackAccess);
++   }
++
++   private static boolean getConfigValue(java.util.function.Supplier<Boolean> configValue) {
++      //Yes we catch the early loading error on purpose. This feature needs to be configurable but Mojang loads resources really early.
++      //So when they are loaded early we preserve the default behaviour which is to process it on the main thread.
++      try {
++         return configValue.get();
++      } catch (IllegalStateException e) {
++         return false;
++      }
++   }
++
+    private static final Map<PackType, Path> f_182296_ = Util.m_137537_(() -> {
+       synchronized(VanillaPackResources.class) {
+          ImmutableMap.Builder<PackType, Path> builder = ImmutableMap.builder();
+@@ -92,6 +_,19 @@
+       this.f_10314_ = ImmutableSet.copyOf(p_143762_);
+    }
+ 
++   @Override
++   public void initForNamespace(final String nameSpace) {
++      if (!shouldUseCache()) return;
++
++      this.cacheManager.index(nameSpace);
++   }
++
++   @Override
++   public void init(final PackType packType) {
++      initForNamespace("minecraft"); //Default namespaces.
++      initForNamespace("realms");
++   }
++
+    public InputStream m_5542_(String p_10358_) throws IOException {
+       if (!p_10358_.contains("/") && !p_10358_.contains("\\")) {
+          if (f_10312_ != null) {
+@@ -146,7 +_,9 @@
+ 
+       try {
+          Path path = f_182296_.get(p_215346_);
+-         if (path != null) {
++         if (shouldUseCache() && path != null && this.cacheManager.hasCached(p_215346_, p_215347_)) {
++            set.addAll(this.cacheManager.getResources(p_215346_, p_215347_, path.getFileSystem().getPath(p_215348_), p_215349_));
++         } else if (path != null) {
```

Comment:
That is really ugly and a lot harder to understand.
I personally think that one extra line shows way better what is happening in this patch file and later in the decompiled source.

Reply:
Actually on second thought, opinion on instead of having the patched in if statement also check `path != null` to just put the patched in chunk inside the existing vanilla if statement? And then just have a simple `else` without brackets for whether vanilla's method should run?